chain/chain/src/types.rs

```diff
@@ -557,19 +557,35 @@ pub trait RuntimeAdapter: Send + Sync {
     ) -> Result<Ordering, Error>;
 
     /// Build receipts hashes.
-    fn build_receipts_hashes(&self, receipts: &[Receipt]) -> Vec<CryptoHash> {
-        let mut receipts_hashes = vec![];
-        for shard_id in 0..self.num_shards() {
-            // importance to save the same order while filtering
-            let shard_receipts: Vec<Receipt> = receipts
-                .iter()
-                .filter(|&receipt| self.account_id_to_shard_id(&receipt.receiver_id) == shard_id)
-                .cloned()
-                .collect();
-            receipts_hashes
-                .push(hash(&ReceiptList(shard_id, &shard_receipts).try_to_vec().unwrap()));
+    // Due to borsh serialization constraints, we have to use `&Vec<Receipt>` instead of `&[Receipt]`
+    // here.
+    fn build_receipts_hashes(&self, receipts: &Vec<Receipt>) -> Vec<CryptoHash> {
+        if self.num_shards() == 1 {
+            return vec![hash(&ReceiptList(0, receipts).try_to_vec().unwrap())];
         }
-        receipts_hashes
+        let mut account_id_to_shard_id = HashMap::new();
```

Comment:
This memoization only helps if there are multiple receipts from the same account in the given `receipts` collection. Do we expect that to happen often? If not, I wonder if the overhead from the memoization would be worse than re-computing a hash every once in a while.

Reply:
The memoization doesn't seem to add much overhead, and does provide significant improvement when there is some duplicates, so I decided it is good to keep.